File: 표/8_graph/순위.py
"""
n명의 권투선수가 권투 대회에 참여, 권투 경기는 1대1 방식
만약 A 선수가 B 선수보다 실력이 좋다면 A 선수는 B 선수를 항상 이깁니다.
몇몇 경기 결과를 분실하여 정확하게 순위를 매길 수 없습니다.
- 선수의 수: n
- 경기 결과를 담은 2차원 배열: results
- [A, B]는 A 선수가 B 선수를 이겼다는 의미
목표: 정확하게 순위를 매길 수 있는 선수의 수를 return

! A가 B에게 이기고 B가 C에게 이기면, A가 C에게 이김
"""

def solution(n, results):
    answer = 0
    
    graph = [[set(), set()] for _ in range(n+1)]
    for win, lose in results:
        graph[win][0].add(lose)
        graph[lose][1].add(win)
        
    for i in range(1, n+1):
        win_list, lose_list = graph[i]
        for w in win_list: # i한테 진 애들은 i를 이긴 애들한테 짐
            graph[w][1].update(graph[i][1])
        for l in lose_list: # i한테 이긴 애들은 i한테 진 애들한테 이김
            graph[l][0].update(graph[i][0])
    
    for i in range(1, n+1):
        if len(graph[i][0]) + len(graph[i][1]) == n-1:
            answer += 1

    return answer

File: 최윤경/순위.js
/*
[접근]

1. 선수 간 승패는 A가 B를 이겼다는 방향이 있는 관계 → 방향 그래프로 표현
2. 직접 경기 결과만으로는 순위를 알 수 없으므로 연결된 승패 관계까지 확인 필요
   → A가 B를 이기고 B가 C를 이기면 A가 C를 이김
3. 모든 선수 사이의 간접 승패 관계를 확인해야 함 → 플로이드-워셜
4. 한 선수가 자신을 제외한 모든 선수와 승패 관계가 정해지면 정확한 순위 확정
5. 승패 관계를 알 수 있는 선수가 n-1명인 선수의 수 반환
*/

function solution(n, results) {
  let answer = 0;
  const win = Array.from({ length: n + 1 }, () => Array(n + 1).fill(false));

  // 승패 저장
  for (const [winner, loser] of results) {
    win[winner][loser] = true;
    win[loser][winner] = false;
  }

  // 연관 승패 저장
  for (let k = 1; k <= n; k++) {
    for (let i = 1; i <= n; i++) {
      for (let j = 1; j <= n; j++) {
        if (win[i][k] && win[k][j]) {
          win[i][j] = true;
        }
      }
    }
  }

  // 선수별 순위 확정 가능 여부 확인
  for (let i = 1; i <= n; i++) {
    let known = 0;

    for (let j = 1; j <= n; j++) {
      if (i === j) continue;

      if (win[i][j] || win[j][i]) {
        known++;
      }
    }

    // 모든 관계 안다면 정답++
    if (known === n - 1) {
      answer++;
    }
  }

  return answer;
}
